### PR TITLE
Add reason, total_shards, include_global_state columns to sys.snapshots

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Relates to https://github.com/crate/crate/pull/16248#discussion_r1661060846

### CODE REVIEW

File: docs/admin/system-information.rst [hunk 1]
```diff
@@ -1956,57 +1956,67 @@ Snapshots
...
++--------------------------+----------------------------------+------------------------------+
+| ``reason``               | Snapshot failure reason.         | ``TEXT``                     |
+|                          | If taking the snapshot was       |                              |
+|                          | successful this is ``NULL``.     |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``total_shards``         | Total number of shards in the    | ``INTEGER``                  |
+|                          | snapshot.                        |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``include_global_state`` | Flag indicating that snapshot    | ``BOOLEAN``                  |
+|                          | was taken with ``ALL`` option.   |                              |
```
Comment (BaurzhanSakhariev) [inline docs/admin/system-information.rst:2018]: Max width changed because of this column, hence adjustments in the first column length for the whole table.

File: docs/appendices/release-notes/5.9.0.rst [hunk 2]
```diff
@@ -106,5 +106,5 @@ Administration and Operations
 - Added support for :ref:`Shared Access Signatures (SAS) tokens <sql-create-repo-azure-sas-token>`
   as an alternative for authentication for :ref:`Azure repositories <sql-create-repo-azure>`.
 
-- Added ``id`` column to the :ref:`sys.snapshots <sys-snapshots>` table which
-  exposes snapshot UUID.
+- Added ``id``, ``reason``, ``total_shards`` and ``include_global_state``
+  columns to the :ref:`sys.snapshots <sys-snapshots>` table.
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/5.9.0.rst:110]: I only kept column names in change entry. I think users can check referenced page to check details, otherwise would have to copy the same info and make the change entry bloated.

File: docs/admin/system-information.rst [hunk 3]
```diff
@@ -1956,57 +1956,67 @@ Snapshots
...
+|                          | while taking the snapshot.       |                              |
+|                          | If taking the snapshot was       |                              |
+|                          | successful this is empty.        |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``reason``               | Snapshot failure reason.         | ``TEXT``                     |
+|                          | If taking the snapshot was       |                              |
+|                          | successful this is ``NULL``.     |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``total_shards``         | Total number of shards in the    | ``INTEGER``                  |
+|                          | snapshot.                        |                              |
```
Comment (BaurzhanSakhariev) [inline docs/admin/system-information.rst:2015]: While debugging flaky test (see second commit fixing it), I realized that this is actualy total _primary_ shards. 

No mention of replicas in `SnapshotInfo` docs

```
     * Returns total number of shards that were snapshotted; a value of {@code 0} will
     * be returned if {@link #state()} returns {@code null}.
```
Should I explicitly mention `primary` here?

Comment (BaurzhanSakhariev) [reply docs/admin/system-information.rst:2015]: > I realized that this is actualy total primary shards.

i.e I checked `sys.shards` - when data nodes is > 1 we have 4 primaries and 4K replicas but totalShards is 4.

Comment (seut) [reply docs/admin/system-information.rst:2015]: > Should I explicitly mention primary here?

Probably. A snapshot never contains replicas as these are just copies of the primaries. Anyway, mentioning it explicitly is worth doing.

File: docs/admin/system-information.rst [hunk 4]
```diff
@@ -1956,57 +1956,67 @@ Snapshots
...
+|                          | snapshot was created with.       |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``failures``             | A list of failures that occurred | ``ARRAY(TEXT)``              |
+|                          | while taking the snapshot.       |                              |
+|                          | If taking the snapshot was       |                              |
+|                          | successful this is empty.        |                              |
++--------------------------+----------------------------------+------------------------------+
+| ``reason``               | Snapshot failure reason.         | ``TEXT``                     |
+|                          | If taking the snapshot was       |                              |
+|                          | successful this is ``NULL``.     |                              |
```
Comment (seut) [inline docs/admin/system-information.rst:2012]:
```suggestion
|                          | If the snapshot was              |                              |
|                          | successful this is ``NULL``.     |                              |
```

File: docs/admin/system-information.rst (hunk 1 above)
Comment (seut) [inline docs/admin/system-information.rst:2018]:
```suggestion
| ``include_global_state`` | Flag indicating that the snapshot    | ``BOOLEAN``                  |
|                          | was taken with the ``ALL`` option.   |                              |
```

File: server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java [hunk 5]
```diff
@@ -105,25 +102,34 @@ public void testUnavailableSnapshotsAreFilteredOut() throws Exception {
...
+            // Basic ctor used in tests, initializes total shards with 0
+            assertThat(sysSnapshot.totalShards()).isEqualTo(0);
+            globalStateFlags.add(sysSnapshot.includeGlobalState());
         });
-        assertThat(names).containsExactlyInAnyOrder("s1", "s2", "s3");
-        assertThat(uuids).containsExactlyInAnyOrder(s1.getUUID(), s2.getUUID(), s3.getUUID());
+        assertThat(names).containsExactly("s1", "s2", "s3");
+        assertThat(uuids).containsExactly(s1.getUUID(), s2.getUUID(), s3.getUUID());
+        // s1 and s2 fail on info retrieval and get undefined/null value for the flag.
+        assertThat(globalStateFlags).containsExactly(null, null, false);
```
Comment (seut) [inline server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java:132]: This test case verifies that "unavailable" snapshots are filtered out. So please do not add any other test scenarios to it but create a dedicated test instead.

Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java]: Good point, I reverted changes here and created a dedicated test

File: server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java [hunk 6]
```diff
@@ -126,6 +127,83 @@ public void testUnavailableSnapshotsAreFilteredOut() throws Exception {
...
+        Metadata metadata = mock(Metadata.class);
+        when(metadata.templates()).thenReturn(ImmutableOpenMap.of());
+
+        // s1 fails for SnapshotInfo (to verify NULL value for includeGlobalState flag).
+        doReturn(CompletableFuture.completedFuture(metadata))
+            .when(r1)
+            .getSnapshotGlobalMetadata(sid1);
+        doReturn(CompletableFuture.failedFuture(new SnapshotException("repo1", "s1", "Everything is wrong")))
+            .when(r1)
+            .getSnapshotInfo(sid1);
```
Comment (seut) [inline server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java:157]: I don't think we should use the error handling at `SysSnapshots.createSysSnapshot()` to create an instance which includes a `null` for the `includeGlobalState`. This is not really related, would be a different test case in my opinion.
Why not just return a instance of `SnapshotInfo` instead like below?

Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java]: Pushed a dedicated test for unavailable snapshot and updated existing test.

Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/reference/sys/snapshot/SysSnapshotsTest.java]: I actually think I should keep only 1 instance of `SnapshotInfo` (with non-default values of new columns) in `test_global_state_and_reason_and_total_shards_are_exposed_in_sys_snapshots`.

There is no actual prod code returning 0 total shards and null flag for successful snapshot.

File: server/src/test/java/io/crate/integrationtests/SysSnapshotsTest.java [hunk 7]
```diff
@@ -62,7 +64,7 @@ protected Settings nodeSettings(int nodeOrdinal) {
     @Test
     @SuppressWarnings("unchecked")
     public void testQueryAllColumns() throws Exception {
-        execute("create table tbl (id int primary key) ");
+        execute("create table tbl (id int primary key) clustered by (id) into 4 shards");
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/SysSnapshotsTest.java:67]: I guess you want to explicitly ensure that there are 4 shards. This is the default, but setting it explicitly as it is asserted may be better. But why the `clustered by(id)` change? Nothing in this test will check for the routing column afaik. Also routing by the primary key is the default.
```suggestion
        execute("create table tbl (id int primary key) clustered into 4 shards;
```

**Review by seut — APPROVED** [review]
LGTM otherwise.

File: server/src/test/java/io/crate/integrationtests/SysSnapshotsTest.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/SysSnapshotsTest.java:67]: >This is the default

it was sometimes different form 4 because of random number of data nodes

Comment (seut) [reply server/src/test/java/io/crate/integrationtests/SysSnapshotsTest.java]: jfyi: number of data nodes aren't related to number of shards ;)
